src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -166,13 +211,43 @@ private Response createSession(Command command) throws IOException {
         setCommandCodec(new AppiumW3CHttpCommandCodec());
         refreshAdditionalCommands();
         setResponseCodec(dialect.getResponseCodec());
-        return result.createResponse();
+        Response response = result.createResponse();
+
+        setDirectConnect(response);
+
+        return response;
     }
 
     public void refreshAdditionalCommands() {
         getAdditionalCommands().forEach(this::defineCommand);
     }
 
+    @SuppressWarnings("unchecked")
+    private void setDirectConnect(Response response) throws MalformedURLException {
+        if (!this.appiumClientConfig.isDirectConnectEnabled()) {
+            return;
+        }
+
+        Map<String, ?> responseValue = (Map<String, ?>) response.getValue();
+
+        String directConnectProtocol = getDirectConnectValue(responseValue, "directConnectProtocol");
```

Comment:
I assume such feature could be a huge security hole in case of non-https connections. Please convince me that I'm wrong.

Also, is it specced officially somewhere?

Reply:
It could be. We can restrict it to accept only HTTPS, but basically, the response depends on the server side's implementation.

Maybe not explicitly. Ordinally this was experimentally added as https://appiumpro.com/editions/86-connecting-directly-to-appium-hosts-in-distributed-environments in some clients like https://github.com/appium/python-client/pull/338 a couple of years ago. So far, Java and C# clients haven't had this. It seems like this was used than a similar experimental execute_driver, btw.

Reply:
lets start by only accepting https as protocol